lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
-          style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
+          style = Nodes::Scalar::LITERAL
+        elsif o.length > line_width
```

Comment:
This is quite odd… why not to simply cache the value in an instance variable? It will be IMHO even better for performance than calling lambda for every string.

```ruby
def line_width
  @line_width ||= (@options[:line_width] || Float::INFINITY)
end
```